api/src/test/java/org/openmrs/api/ProgramWorkflowServiceTest.java

```diff
@@ -740,7 +750,263 @@ public void shouldTestPurgeProgramAttributeType() throws Exception {
                 int totalAttributeTypes = pws.getAllProgramAttributeTypes().size();
 		pws.purgeProgramAttributeType(programAttributeType);
 		assertEquals((totalAttributeTypes - 1), pws.getAllProgramAttributeTypes().size());
-}
+	}
+	
+	
+	@Test
+	public void getPrograms_shouldTestGetPrograms() {
+		List<Program> malPrograms = pws.getPrograms("MAL");
+		List<Program> prPrograms = pws.getPrograms("PR");
+		assertEquals(malPrograms.size(), 1);
+		assertEquals(prPrograms.size(), 3);
+	}
+	
+	@Test
+	public void retireProgram_shouldSetRetiredStateToFalseAndSetAReason() {
+		Program programTest = pws.getAllPrograms().get(0);
+		Program program = pws.retireProgram(programTest, "expired");
+		for (ProgramWorkflow workflow : program.getWorkflows()) {
+			assertEquals(workflow.getRetired(), true);
+			for (ProgramWorkflowState state : workflow.getStates()) {
+				assertNotNull(workflow.getStates());
+				assertEquals(state.getRetired(), true);
+				assertEquals(state.getRetired(), true);
+			}
+		}
+		assertEquals(program.getRetireReason(), "expired");
+		
+	}
+	
+	@Test
+	public void unretireProgram_shouldSetRetireFalseForWorkflowsAndWorkflowStates() {
+		Program program = pws.getAllPrograms().get(0);
+		Date lastModifiedDate = program.getDateChanged();
+		Program program1 = pws.unretireProgram(program);
+		assertEquals(program.getRetired(), false);
+		for (ProgramWorkflow workflow : program.getAllWorkflows()) {
+			if (lastModifiedDate != null && lastModifiedDate.equals(workflow.getDateChanged())) {
+				assertEquals(workflow.getRetired(), false);
+				for (ProgramWorkflowState state : workflow.getStates()) {
+					if (lastModifiedDate.equals(state.getDateChanged())) {
+						assertEquals(state.getRetired(), false);
+					}
+				}
+			}
+		}
+	}
+	
+	@Test
+	public void saveUnretireProgram_shouldTestSaveUnretireProgram() {
+		Program program = new Program();
+		Concept concept = Context.getConceptService().getAllConcepts().get(0);
+		User testUser = Context.getUserService().getAllUsers().get(0);
+		program.setConcept(concept);
+		program.setDescription("test");
+		program.setName("programTest");
+		program.setCreator(testUser);
+		pws.getAllPrograms();
+		pws.retireProgram(program, "abc");
+		pws.unretireProgram(program);
+		Program program1 = dao.getProgramByUuid(program.getUuid());
+		assertEquals(program, program1);
+	}
+	
+	@Rule
+	public ExpectedException expectedException = ExpectedException.none();
```

Comment:
This should be at the top of the file with all the other fields. Also, it should be marked as `protected`.

Reply:
Tried to do this and caused the following error:
https://pastebin.com/bJnE0QDu